server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -761,7 +794,7 @@ public synchronized void reset() {
         public synchronized void addFileDetail(String name, long length, boolean reused) {
             assert fileDetailsComplete == false : "addFileDetail for [" + name + "] when file details are already complete";
             File file = new File(name, length, reused);
-            File existing = fileDetails.put(name, file);
+            File existing = fileDetails.addFileDetails(name, file);
             assert existing == null : "file [" + name + "] is already reported";
```

Comment:
This assertion feels like it belongs in `RecoveryFileDetails` too, as does the one that asserts `fileDetailsComplete == false`.